### ISSUE TITLE
fix(public-ip): disassociate from resources before destroy

### ISSUE THREAD

**Issue by hknutsen** [equinor/terraform-azurerm-network#37]
When running the complete example, I get the following error:

```
Returning due to fatal error: FatalError{Underlying: error while running command: exit status 1; ╷
│ Error: deleting Public Ip Address: (Name "pip-a969dea59fa41b71" / Resource Group "rg-a969dea59fa41b71"): network.PublicIPAddressesClient#Delete: Failure sending request: StatusCode=400 -- Original Error: Code="PublicIPAddressCannotBeDeleted" Message="Public IP address /subscriptions/ce8c6461-a20d-47a3-884c-6659cb74e53f/resourceGroups/rg-a969dea59fa41b71/providers/Microsoft.Network/publicIPAddresses/pip-a969dea59fa41b71 can not be deleted since it is still allocated to resource /subscriptions/ce8c6461-a20d-47a3-884c-6659cb74e53f/resourceGroups/rg-a969dea59fa41b71/providers/Microsoft.Network/natGateways/ng-a969dea59fa41b71. In order to delete the public IP, disassociate/detach the Public IP address from the resource.  To learn how to do this, see aka.ms/deletepublicip." Details=[]
│ 
│ 
╵}
    /home/henrik/gitrepos/equinor/terraform-azurerm-network/test/destroy.go:11:
                Error Trace:    destroy.go:11
                                                        complete_example_test.go:17
                Error:          Received unexpected error:
                                FatalError{Underlying: error while running command: exit status 1; ╷
                                │ Error: deleting Public Ip Address: (Name "pip-a969dea59fa41b71" / Resource Group "rg-a969dea59fa41b71"): network.PublicIPAddressesClient#Delete: Failure sending request: StatusCode=400 -- Original Error: Code="PublicIPAddressCannotBeDeleted" Message="Public IP address /subscriptions/ce8c6461-a20d-47a3-884c-6659cb74e53f/resourceGroups/rg-a969dea59fa41b71/providers/Microsoft.Network/publicIPAddresses/pip-a969dea59fa41b71 can not be deleted since it is still allocated to resource /subscriptions/ce8c6461-a20d-47a3-884c-6659cb74e53f/resourceGroups/rg-a969dea59fa41b71/providers/Microsoft.Network/natGateways/ng-a969dea59fa41b71. In order to delete the public IP, disassociate/detach the Public IP address from the resource.  To learn how to do this, see aka.ms/deletepublicip." Details=[]
                                │ 
                                │ 
                                ╵}
```

According to [docs](https://registry.terraform.io/providers/hashicorp/azurerm/3.59.0/docs/resources/public_ip) it seems like a possible solution would be to add the following to the public IP resource:

```
  lifecycle {
    create_before_destroy = true
  }
```

**Comment by hknutsen** [comment]
@kov117 Tests still pass for me on v1.13 (`git checkout v1.13.0` then run tests).

Maybe look into the changes to made to the examples for v1.14?

**Comment by kov117** [comment]
the changes are the added feature of the sub-module `nic` and a resource "azurerm_public_ip" "example", and the error messages points at the public ip resource.

```
 Error:          Received unexpected error:
                                FatalError{Underlying: error while running command: exit status 1; ╷
                                │ Error: waiting for creation/update of Public Ip Address: (Name "pip-371143fd4a677833" / Resource Group "rg-371143fd4a677833"): Code="Canceled" Message="Operation was canceled." Details=[{"code":"CanceledAndSupersededDueToAnotherOperation","message":"Operation PutPublicIpAddressOperation (27faa138-8a5f-4000-84d4-c038621d0881) was canceled and superseded by operation PutPublicIpAddressOperation (62e9d86d-2603-4517-9191-cb6b4085e214)."}]
                                │
                                │   with azurerm_public_ip.example,
                                │   on main.tf line 163, in resource "azurerm_public_ip" "example":
                                │  163: resource "azurerm_public_ip" "example" {
                                │
                                ╵}
                Test:           TestCompleteExample
```

could the issue have something to do with there being a module public ip and a resource public ip ?

**Comment by kov117** [comment]
got a test passed when i removed the resource public ip, so the error is definetely something involved with that resource, but i am not sure how to fix the error without removing the public ip resource, any suggestions ?

**Comment by hknutsen** [comment]
Have you checked if tests for v1.13 pass on your side? If yes, then the error isn't necessarily with the Public IP resource, but with how that resource is used in the examples in v1.14. If so, investigate what changes need to be made to those example.

**Comment by kov117** [comment]
yes the test passed for v1.13, i will investigate your suggestions 👍🏽

**Comment by kov117** [comment]
i believe i have found the issue, there might have been a mix up when i merged `nic` branch to the main (#30) i added the resource public ip when originally it was supposed to be replaced with the module public ip. if you take a look at the closed issue #34, and then look at the files changed tab, in the basic complete, you will see that the resource public ip is being replaced with the module public ip.

**Comment by kov117** [comment]
i have no clue as to why or how i got both resource and module public ip when i merged the branch to main, but so far i am quite certain that this is the cause of the issue.

**Comment by kov117** [comment]
this issue will be fixed along with #33

**Comment by hknutsen** [comment]
I think a separate PR for fixing this is better.

To be fixed:
- Remove duplicate public IPs: https://github.com/equinor/terraform-azurerm-network/blob/3dee6975f5f463354e044f2533322d978790ed25/examples/complete/main.tf#L163-L183
- NIC should not be connected to hub network: https://github.com/equinor/terraform-azurerm-network/blob/3dee6975f5f463354e044f2533322d978790ed25/examples/complete/main.tf#L155

**Comment by hknutsen** [comment]
Issue was that we were creating multiple public IPs of the same name in the complete exampled.

Fixed in #41.
